Examination: Converter

SunnyWorker opened the pull request.

SunnyWorker committed:
OperationType enum added

SunnyWorker committed:
Currency and ExchangeCurrency classes added

SunnyWorker committed:
Loading exchange rates from config file added

SunnyWorker committed:
Currency and OperationType classes functionality extended
IllegalOperationException added

SunnyWorker committed:
Currency and OperationType classes functionality extended
Tests modified

SunnyWorker committed:
Library user code for Main created

SunnyWorker committed:
Converter realisation modified

Versach commented on src/main/java/org/example/Main.java:
```diff
@@ -1,7 +1,17 @@
+        Scanner sc = new Scanner(System.in);
+        IConverter converter = new Converter();
+        while(true) {
+            System.out.println("Enter request: ");
```
было бы неплохо вынести в константы либо в в проперти эти "приветственные" сообщения. Облегчает потенциальную локализацию и снижает необходимость в редактировании каких то вещей при тестировании

Versach commented on src/main/java/org/example/converter/enums/OperationType.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.regex.Pattern;
+
+public enum OperationType {
+    toDollars(2, Pattern.compile("toDollars"+Currency.getPatternOfAllCurrencies())),
+    toRubles(2, Pattern.compile("toRubles"+Currency.getPatternOfAllCurrencies())),
+    add(1,Pattern.compile(Currency.getPatternOfAllCurrencies()+"\\+"+Currency.getPatternOfAllCurrencies())),
+    subtract(1, Pattern.compile(Currency.getPatternOfAllCurrencies()+"-"+Currency.getPatternOfAllCurrencies()));
+
```
обычно "плюс" от окружения отделен пробелами. Сами команды(toDollars и тп) тоже просятся в константы

Versach commented on src/main/java/org/example/converter/implementations/Converter.java:
```diff
@@ -0,0 +1,95 @@
+        for (int i = 0; i < request.length(); i++) {
+            if(request.charAt(i)=='(') brackets++;
+            if(request.charAt(i)==')') brackets--;
+            if(brackets<0)
+                throw new RequestSyntaxException("Incorrect using of brackets in request!");
+        }
+        if(brackets!=0)
+            throw new RequestSyntaxException("Incorrect using of brackets in request!");
+    }
+
```
не хватает пробелов рядом с операторами сравнения. Используй ctrl+alt+l

Versach commented on src/main/java/org/example/dao/IExchangeRateDAO.java:
```diff
@@ -0,0 +1,10 @@
+
+import java.util.HashMap;
+
+public interface IExchangeRateDAO {
```
А почему это DAO?

Versach commented on src/test/java/org/example/ExchangeCalculatorTest.java:
```diff
@@ -18,7 +25,7 @@ public void toRublesSimpleOperationTest() {
     @Test
     public void toDollarsSimpleOperationTest() {
         assertEquals("$1,00", converter.performOperation("toDollars(  60,00р)"));
-        assertEquals("$1,50", converter.performOperation(" toDollars(90,00р  )"));
+        assertEquals("$1,49", converter.performOperation(" toDollars(90,00р  )"));
     }
```
nit: снова чехарда с пробелами. Могу так же посоветовать использовать всякие статические анализаторы кода потипу sonar или checkstyle

Versach commented on src/test/java/org/example/ExchangeCalculatorTest.java:
```diff
@@ -44,13 +51,13 @@ public void rublesExpressionOperationsTest() {
         assertEquals("$0,50", converter.performOperation("$1,50 + $1,50 - $2,50"));
-        assertEquals("$-0,50", converter.performOperation("$1,50  -  toDollars(120,00р)"));
+        assertEquals("$-0,49", converter.performOperation("$1,50  -  toDollars(120,00р)"));
     }
 
     @Test
     public void innerOperationsTest() {
-        assertEquals("$4,00", converter.performOperation("toDollars(120,00р + 120,00р)"));
-        assertEquals("$1,50", converter.performOperation("toDollars(toRubles($1,50))"));
-        assertEquals("$4,00", converter.performOperation("toDollars(toRubles($1,50  + $2,50))"));
+        assertEquals("$3,98", converter.performOperation("toDollars(120,00р + 120,00р)"));
+        assertEquals("$1,49", converter.performOperation("toDollars(toRubles($1,50))"));
+        assertEquals("$3,98", converter.performOperation("toDollars(toRubles($1,50  + $2,50))"));
```
в целом так многие делают, но подгонка тестов под результаты на самом деле это плохой тест( хотя иногда и единственно возможный =).

Одним из самых верных подходов является так называемая разработка через тестирование, когда ты сперва пишешь тест, точно зная что должно быть в результате. Создаешь метод без кода  под операцию сложения и тупо ноль всегда возвращаешь например
Затем пишешь тест исходя из нее. Тесть пытаемся выполнить условные 2 + 2 и проверяем что оно равно 4. Мы заранее считаем и пишем в тест заведомо проверенную информацию, иными словами - тест сразу пишем хорошо и верно, а затем уже код дорабатываем пока он не начнет проходить тест, а не наоборот

ты в целом как будто бы так и начал если посмотреть на мастер ветку, но почему то потому переписал тесты, возможно ты просто что то изменил относительно изначальной задумки.

Versach reviewed (commented):
.idea и прочие файлы среды разработки обычно не включают в коммиты
С большего все ок, но стоит повыносить в константы некоторые вещи

SunnyWorker committed:
Correcting cons
Add .idea folder to gitignore
Refactoring ExchangeRateDAO to ExchangeRateImporter
Refactoring Main to Launcher
Improving code style
Constants of operations add to OperationType
Constants of console messages add to Launcher

Versach commented on src/main/java/org/example/converter/enums/Currency.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.regex.Pattern;
+
+public enum Currency {
+    dollarUSA("\\$-?\\d+(,\\d+)?"), russianRuble("-?\\d+(,\\d+)?р");
```
Енамы обычно как константы записывают ( используя upper case и underscope)

Versach commented on src/main/java/org/example/converter/enums/OperationType.java:
```diff
@@ -0,0 +1,29 @@
+
+public enum OperationType {
+    toDollars(2, Pattern.compile("toDollars" + Currency.getPatternOfAllCurrencies())),
+    toRubles(2, Pattern.compile("toRubles" + Currency.getPatternOfAllCurrencies())),
```
А почему тут не константы?

Versach commented on src/main/java/org/example/converter/implementations/Converter.java:
```diff
@@ -0,0 +1,94 @@
+            }
+        }
+        if (!Pattern.compile(Currency.getPatternOfAllCurrencies()).matcher(expression).matches())
+            throw new RequestSyntaxException("String " + expression + " doesn't match any possible pattern!");
```
лучше использовать formatter и строку - шаблон вынести в константу. можно даже в отдельный файл если еще какие то тексты ошибок подобные есть

Versach commented on src/main/java/org/example/converter/implementations/OperationExecutor.java:
```diff
@@ -0,0 +1,106 @@
+            Matcher matcher = pattern.matcher(operationString);
+            if (matcher.find()) {
+                if (value.equals(currency))
+                    throw new IllegalOperationException("You can't transform currencies into themselves!");
```
А ну вот типо такой ошибки да, можно все тексты ошибок в отдельный файл с константами

Versach commented on src/main/java/org/example/dao/ExchangeRateFileSystemImporter.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Scanner;
+
+public class ExchangeRateFileSystemImporter implements IExchangeRateImporter {
+    private static final String configurationFilename = "exchange_rates.bank";
```
Поля-константы именуются с использованием upper case и underscope

Versach commented on src/test/java/org/example/ConverterTest.java:
```diff
@@ -38,19 +46,19 @@ public void dollarsSummarySimpleOperationTest() {
-        assertEquals("$4,00", converter.performOperation("toDollars(120,00р + 120,00р)"));
-        assertEquals("$1,50", converter.performOperation("toDollars(toRubles($1,50))"));
-        assertEquals("$4,00", converter.performOperation("toDollars(toRubles($1,50  + $2,50))"));
+        assertEquals("$3,98", converter.performOperation(OperationType.toDollarsOperation + "(120,00р + 120,00р)"));
```
если импортировать операции с помощью static import можно будет не писать OperationType-ы

Versach reviewed (commented):
можно еще доработать

SunnyWorker committed:
Final improvement
Constants code convention corrected
Derived exceptions created
Configuration file style corrected
import static added to unit tests
Package dao renamed to importer

Versach commented on src/main/java/org/example/Launcher.java:
```diff
@@ -0,0 +1,20 @@
+        Scanner sc = new Scanner(System.in);
+        IConverter converter = new Converter();
+        while (true) {
+            System.out.println(ENTRY_MESSAGE);
```
Nit: "entry" тут неоч подходит

Versach commented on src/main/java/org/example/converter/implementations/Converter.java:
```diff
@@ -0,0 +1,95 @@
+                }
+            }
+        }
+        if (!Pattern.compile(Currency.getPatternOfAllCurrencies()).matcher(expression).matches())
```
Nit:  желательно стремится избегать отрицаний, тяжело читаются

Versach reviewed (commented):
теперь ок
